specs/networking/libp2p-standardization.md

```diff
@@ -135,24 +161,369 @@ number of shard subnets is defined via `SHARD_SUBNET_COUNT` and the shard
+libp2p-specific modifications - as suggested by @Zah.*
+
+
+### Specification
+
+#### Protocol Segregation
+
+Each RPC message is segregated into it's own libp2p protocol id, which is a string of the form:
+
+```
+<ProtocolPrefix>/<MessageName>/<SchemaVersion>/<Encoding>
+```
+
+With:
+* **ProtocolPrefix** -- the RPC messages are grouped into families identified
+  by a shared LibP2P protocol name prefix. A conforming implementation is
+  expected to support either all messages appearing within a family or none of
+  them. In this case, we use `/eth/serenity/rpc`.
+* **MessageName** -- each RPC request is identified by a name consisting of
+English letters, digits and underscores (_).
+* **SchemaVersion** -- a semantic version consisting of one or more numbers
+  separated by dots (.). Each schema is versioned to facilitate backward and
+  forward-compatibility when possible.
+* **Encoding** -- while the schema defines the data types in more abstract
+  terms, the encoding describes a specific representation of bytes that will be
+  transmitted over the wire. See the [Encodings](#encodings) section, for
+  further details.
+
+
+This protocol segregation allows libp2p `multistream-select` to handle the
+RPC-type, version and encoding negotiation before establishing the underlying
+substreams.
+
+
+#### Requests and Responses
+
+Each request/response has a maximum byte size of `RPC_MAX_SIZE`. This exact
+value is TBD.
+
+Requests/Responses MUST be encoded such that a prefixed length (defined by the
+encoding used) is provided. Clients SHOULD ensure the request/response size is
+less than or equal to `RPC_MAX_SIZE`, if not, SHOULD reset the substream. 
+
+A receiver SHOULD decode the length-prefix and wait for the exact number of
+bytes to be sent. Once received the requester closes the substream.
+
+##### Request
+
+A request is formed by initiating a connection with the protocol id matching
+the desired request type, encoding and version. Once a successful substream is
+negotiated, the request is sent with the matching encoding (see the
+[Encoding](#encoding) section for further details).
+
+Once a stream is negotiated, the requester SHOULD send the request within **3
+seconds**.
+
+The requester SHOULD then wait for a response on the negotiated stream for at
+most **10 seconds**, before resetting the stream. 
+
+*Note: If a request does not require a response, such as with a `Goodbye`
+message, the stream is closed instantly.*
+
+##### Response
+
+After a request has been received on a negotiated stream, the responder SHOULD
+send a response within **10 seconds**.  A response consists of a **single-byte**
+response code which determines the contents of the response.
+
+It can have one of the following values
+
+* 0: **Success** -- a normal response with contents matching the expected
+  message schema and encoding specified in the request.
+* 1: **EncodingError** -- the receiver was not able to decode and deserialize
```

Comment:
Why not merge this with `InvalidRequest`? I think this would be nice from a conceptual point of view: Either you did something wrong, I did something wrong, or everything is ok. Also, `EncodingError`s should never happen anyways.

Reply:
Yep I agree. I had headaches with implementing this anyway. The code will also be cleaner.

Reply:
I agree, it's reasonable to use `InvalidRequest` in place of `EncodingError`.